### PR TITLE
Add readonlyArray.find macro

### PR DESCRIPTION
**Pull request by Dog2puppy**
TS:
```ts
let myFoundThing = frameworks.find((val) => {
	return true
})
```
produces this in Lua:
```lua
-- ▼ ReadonlyArray.find ▼
local _1 = function(val)
	return true
end
local _2 = nil
for _3, _4 in ipairs(frameworks) do
	if _1(_4, _3 - 1, frameworks) == true then
		_2 = _4
		break
	end
end
-- ▲ ReadonlyArray.find ▲
local myFoundThing = _1
```
Credit to @Dionysusnu for helping me fix type errors and returning what was found within the loop.

### CODE REVIEW

File: src/TSTransformer/macros/propertyCallMacros.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ import { ensureTransformOrder } from "TSTransformer/util/ensureTransformOrder";
 import { isUsedAsStatement } from "TSTransformer/util/isUsedAsStatement";
 import { offset } from "TSTransformer/util/offset";
 import { assert } from "Shared/util/assert";
+import { SyntaxKind } from "typescript";
```
Comment (Dionysusnu) [inline src/TSTransformer/macros/propertyCallMacros.ts:9]: Don't import anything from typescript directly.

Comment (Dionysusnu) [reply src/TSTransformer/macros/propertyCallMacros.ts:9]: If you need it, use `import ts from "byots"`

File: src/TSTransformer/macros/propertyCallMacros.ts [hunk 2]
```diff
@@ -454,7 +455,14 @@ const READONLY_ARRAY_METHODS: MacroList<PropertyCallMacro> = {
 				start: lua.number(1),
 				end: size,
 				step: lua.number(1),
```
Comment (Dionysusnu) [inline src/TSTransformer/macros/propertyCallMacros.ts:457]: Defaults to 1 in lua. Necessary?

File: src/TSTransformer/macros/propertyCallMacros.ts (hunk 1 above)
Comment (Dog2puppy) [reply src/TSTransformer/macros/propertyCallMacros.ts:9]: Wait howd that get added I never touched the imports

File: src/TSTransformer/macros/propertyCallMacros.ts (hunk 2 above)
Comment (Dog2puppy) [reply src/TSTransformer/macros/propertyCallMacros.ts:457]: start and step are both required and the 1 is needed because the argument isn't optional.

Comment (Dionysusnu) [reply src/TSTransformer/macros/propertyCallMacros.ts:456]: Step is not required. I think it looks cleaner to leave it undefined.

File: src/TSTransformer/macros/propertyCallMacros.ts [hunk 3]
```diff
@@ -434,7 +434,58 @@ const READONLY_ARRAY_METHODS: MacroList<PropertyCallMacro> = {
...
+		const callbackId = state.pushToVarIfComplex(transformExpression(state, node.arguments[0]));
+		const loopId = lua.tempId();
+		const valueId = lua.tempId();
+		const returnId = state.pushToVar(lua.nil());
+
+		const size = lua.create(lua.SyntaxKind.UnaryExpression, {
+			operator: "#",
+			expression,
+		});
+
```
Comment (daniel-melanson) [inline src/TSTransformer/macros/propertyCallMacros.ts:450]: `size` is already defined in the global scope. You don't need this here.

File: src/TSTransformer/macros/propertyCallMacros.ts [hunk 4]
```diff
@@ -434,7 +434,58 @@ const READONLY_ARRAY_METHODS: MacroList<PropertyCallMacro> = {
...
+		state.prereq(
+			lua.create(lua.SyntaxKind.NumericForStatement, {
+				id: loopId,
+				start: lua.number(1),
+				end: size,
```
Comment (daniel-melanson) [inline src/TSTransformer/macros/propertyCallMacros.ts:455]: Change `size` to `size(state, node, expression)`

**Comment by osyrisrblx** [comment]
This looks good, but maybe we should use ipairs instead?
In Luau, ipairs is faster than a numeric for-loop (when you also do `arr[i]`)

**Comment by Dionysusnu** [comment]
Wait, really??

**Pull request merged**